### PR TITLE
add ClassName constructors

### LINKED ISSUE #730: please give me a constructor to make ClassName with List<String>
Now I can new `ClassName` with two `String` and vararg `String`s. but I would like constructor with `List<String>`.

Of course, I know that way to make  `ClassName` with `List<String>`. but it is not good for performance:
```
val names = listOf("kotlin.collections", "Map", "Entry")
val className = ClassName(names[0], names[1], *names.drop(2).toTypedArray())
```

### PR DESCRIPTION
**Pull request by anatawa12**
Fixes #730

### CODE REVIEW

**Comment by anatawa12** [comment]
@JakeWharton you seem to be waiting for that I marge this. But I can't marge this because I don't have write access. 

If you aren't waiting, could you tell me why you don't marge?

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeName.kt [hunk 1]
```diff
@@ -288,9 +288,28 @@ class ClassName internal constructor(
    * Returns a class name created from the given parts. For example, calling this with package name
    * `"java.util"` and simple names `"Map"`, `"Entry"` yields `Map.Entry`.
    */
+  @Deprecated("", level = DeprecationLevel.HIDDEN)
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeName.kt]: This is a binary incompatible change that will affect Java callers:
```java
String[] names = { "simple2", "simple3" };
ClassName cn3 = new ClassName("package", "simple", names);
```
This code compiles before the change and won't compile after. Normally we would require a major version change to merge binary-incompatible code, and given that this change is an API enhancement, not a bug fix, I would rather postpone merging this until we're ready for 2.0. @JakeWharton what are your thoughts?

**Comment by ZacSweers** [comment]
Why not give it a different JvmName for Java callers?

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeName.kt (hunk 1 above)
Comment (JakeWharton) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeName.kt]: It's source incompatible for that specific case, not binary. And I highly doubt anyone is using it that way. It's far more common to do `get("package", "Map", "Entry")` which remains source compatible.

I don't have any problem breaking source compatibility for a usability improvement.

**Comment by Egorand** [comment]
@ZacSweers It's a constructor.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeName.kt (hunk 1 above)
Comment (Egorand) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeName.kt]: Yeah, that's fair.

**Comment by anatawa12** [comment]
I signed CLA.

**Comment by anatawa12** [comment]
Thank you for accepting my future request and merging PR.

**Comment by Egorand** [comment]
@anatawa12 thanks for your PR!

**Pull request merged**